### PR TITLE
Implemented new solution for Requests in VR Module

### PR DESCRIPTION
**Pull request by KKarlash**
@bonewell 
@VladSemenyuk 
@DKlimenko 
@giangtqh 
@thinhts

### CODE REVIEW

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+ private:
+  VRModule* parent_;
+  application_manager::MessagePtr json_message_;
+};
+
+}  // namespace commands
+
+}  // namespace vr_cooperation
+#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_BASE_GPB_REQUEST_H_ */
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: `/*` -> ` //`

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+class BaseGpbRequest : public BaseCommandRequest,
+    public event_engine::EventObserver<vr_hmi_api::ServiceMessage,
+        vr_hmi_api::RPCName> {
+ public:
+
+  /**
+   * @brief BaseGpbRequest class constructor
+   * @param parent pointer to VRModule
+   * @param message Message from Mobile
+   **/
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: */

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_json_request.h [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+ */
+class BaseJsonRequest : public BaseCommandRequest,
+    public event_engine::EventObserver<application_manager::MessagePtr,
+        vr_hmi_api::RPCName> {
+ public:
+  /**
+   * @brief BaseJsonRequest class constructor
+   * @param parent pointer to VRModule
+   * @param message Message from HMI
+   **/
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_json_request.h]: */

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h (hunk 2 above)
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: @giangtqh could you please help to find appropriate requirement in coding style?

Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: I do not see this requirement in coding style, but block comments / description should be consistent

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_json_request.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          vr_hmi_api::RPCName>& event) = 0;
+
+  /**
+   * @brief Parse result code from response
+   *
+   * @param value message in response from Mobile
+   * @param result_code Outgoing param with result code
+   */
+  void ParseMobileResultCode(const Json::Value& value,
+                             vr_hmi_api::ResultCode& result_code);
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_json_request.h]: Please, look at [Google style](https://google.github.io/styleguide/cppguide.html#Reference_Arguments)

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_json_request.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+  /**
+   * @brief Prepares request message for Mobile
+   * @param function_id request ID
+   * @param msg_params params from gpb message
+   * @param message that will be sent to mobile
+   */
+  void PrepareRequestMessageForMobile(vr_hmi_api::RPCName function_id,
+                                      const std::string& message_params,
+                                      application_manager::MessagePtr& message);
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_json_request.h]: Don't use reference for smart pointers.

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h (hunk 2 above)
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: It's not Google style. It's Doxygen rules.

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h [hunk 6]
```diff
@@ -104,4 +104,4 @@ class BaseGpbRequest : public BaseCommandRequest,
 }  // namespace commands
 
 }  // namespace vr_cooperation
-#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_BASE_GPB_REQUEST_H_ */
+#endif // SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_COMMANDS_BASE_GPB_REQUEST_H_
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: Double space before `//` ;)

**Comment by bonewell** [comment]
Reviewed.

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+   * @brief BaseGpbRequest on timeout reaction
+   */
+  virtual void OnTimeout();
+
+  /**
+   * @brief This method will be called when receive an event
+   *
+   * @param event The received event
+   */
+  virtual void on_event(
```
Comment (DKlimenko) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: What is the difference between on_event and OnEvent?

File: src/components/vr_cooperation/src/command/base_json_request.cc [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+void BaseJsonRequest::ParseMobileResultCode(
+    const Json::Value& value, vr_hmi_api::ResultCode& result_code) {
+  std::string mobile_result_code;
+  if (IsMember(value, kResult) && IsMember(value[kResult], kCode)) {
+    mobile_result_code = value[kResult][kCode].asString();
+  } else if (IsMember(value, kError) && IsMember(value[kError], kCode)) {
+    mobile_result_code = value[kError][kCode].asString();
+  }
```
Comment (DKlimenko) [inline src/components/vr_cooperation/src/command/base_json_request.cc]: Please add else with error message

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      const event_engine::Event<vr_hmi_api::ServiceMessage, vr_hmi_api::RPCName>& event);
+
+ protected:
+  /**
+   * @brief Interface method that is called whenever new event received
+   */
+  virtual void OnEvent(
```
Comment (DKlimenko) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: May be it is better to rename OnEvent to ProcessEvent with proper description?

**Comment by DKlimenko** [comment]
Reviewed

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h (hunk 7 above)
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: on_event could have additional logic in parent class
OnEvent call child logic

File: src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h (hunk 9 above)
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/commands/base_gpb_request.h]: ok

File: src/components/vr_cooperation/src/command/base_json_request.cc (hunk 8 above)
Comment (KKarlash) [inline src/components/vr_cooperation/src/command/base_json_request.cc]: ok

**Comment by bonewell** [comment]
Reviewed.

**Pull request merged**